src/interrupts/mod.rs

```diff
@@ -0,0 +1,586 @@
+            panic!("Disable interrupts first");
+        }
+    }
+}
+
+/// Creates a new scope, to guarantee that the `InterruptTable` constructor is called.
+///
+/// # Examples
+/// ```rust
+/// fn main(hw: board::Hardware) -> ! {
+///     // Extract hardware
+///     let board::Hardware {
+///         rcc,
+///         pwr,
+///         flash,
+///         nvic,
+///         ..
+///     } = hw;
+///    
+///     // Configure system clock
+///     system_clock::init(rcc, pwr, flash);
+///
+///     use stm32f7::interrupts::interrupt_request::InterruptRequest::Tim7;
+///     use interrupts::Priority::P1;
+///     // Open scope with interrupt support
+///     interrupts::scope(nvic, |irq| { hprintln!("Default handler: {}", irq) },
+///         |interrupt_table| {
+///             let _ = interrupt_table.register(Tim7, P1, || {
+///                 hprintln!("Interrupt handler for Tim7");
+///         });
+///     });
+///    loop{}
+/// }
+/// ```
+///
+/// # Panics
+/// Panics if an interrupt is enabled and is not disabled after use in `code()`
+pub fn scope<'a,F,C,R>(nvic: &'static mut Nvic, default_handler: F, code: C) -> R 
+    where F: FnMut(u8) + 'a,
+        C: FnOnce(&mut InterruptTable<'a>) -> R
+{
+    unsafe {
+        debug_assert!(DEFAULT_HANDLER.is_none());
+        DEFAULT_HANDLER = Some(transmute::<Box<FnMut(u8) + 'a>, Box<FnMut(u8) + 'static>>(Box::new(default_handler)));
+    }
+    
+    let mut interrupt_table = InterruptTable {
+        _lifetime: PhantomData,
+        nvic: nvic,
+        data: [None; 98],
+    };
+    // When the *code(self)* panics, the programm ends in an endless loop with disabled interrupts
+    // and never returns. So the state of the ISRS does't matter.
+    code(&mut interrupt_table)
+
+    // Drop is called
+}
+
+impl<'a> InterruptTable<'a> {
+
+    /// Registers an interrupt with the lifetime of the `InterruptTable`.
+    ///
+    /// # Examples
+    /// ```
+    /// use stm32f7::interrupts::interrupt_request::InterruptRequest::Tim7;
+    /// use interrupts::Priority::P1;    
+    /// interrupts::scope(nvic, |irq| { hprintln!("Default handler: {}", irq) },
+    ///     |interrupt_table| {
+    ///             let interrupt_handle = interrupt_table.register(Tim7, P1,
+    ///             || {
+    ///                 // Isr for interrupt `Tim7`
+    ///             }).expect("Interrupt already used");
+    ///
+    ///             /* Code that needs interrupt `Tim7` */
+    ///
+    ///             // Unregister interrupt and get back the ownership to `data`
+    ///             let data = interrupt_table.unregister(interrupt_handle);
+    ///             assert!(data.is_none());
+    /// });
+    /// ```
+    pub fn register<F>(&mut self,
+                              irq: InterruptRequest,
+                              priority: Priority,
+                              isr: F)
+                              -> Result<InterruptHandle<()>, Error>
+        where F: FnMut() + 'a + Send
```

Comment:
couldn't this function's body simply call `register_owned(irq, priority, (), isr)`?

The `()` will get boxed, but that doesn't even allocate.

That would allow `unregister` to return a value unconditionally, instead of returning an `Option`, which should never be `None` except if called on an `InterruptHandle` created by `register` right now.